File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/live/listener/LrcActionListenerImpl.java
package io.agora.scene.ktv.live.listener;

import android.content.Context;

import io.agora.karaoke_view.v11.model.LyricsLineModel;
import io.agora.karaoke_view.v11.model.LyricsModel;
import io.agora.scene.ktv.live.RoomLivingViewModel;
import io.agora.scene.ktv.widget.lrcView.LrcControlView;

/**
 * 调音台 listener
 */
public class LrcActionListenerImpl implements LrcControlView.OnKaraokeEventListener {

    private final Context mContext;
    private final RoomLivingViewModel mViewModel;
    private final LrcControlView mLrcControlView;

    /**
     * Instantiates a new Lrc action listener.
     *
     * @param context        the context
     * @param viewModel      the view model
     * @param lrcControlView the lrc control view
     */
    public LrcActionListenerImpl(Context context, RoomLivingViewModel viewModel, LrcControlView lrcControlView) {
        mContext = context;
        mViewModel = viewModel;
        mLrcControlView = lrcControlView;
    }


    @Override
    public void onSwitchOriginalClick(LrcControlView.AudioTrack audioTrack) {
        LrcControlView.OnKaraokeEventListener.super.onSwitchOriginalClick(audioTrack);
        mViewModel.musicToggleOriginal(audioTrack);

        if (audioTrack == LrcControlView.AudioTrack.Origin ||
                audioTrack == LrcControlView.AudioTrack.DaoChang) {
            mLrcControlView.setSwitchOriginalChecked(true);
        } else {
            mLrcControlView.setSwitchOriginalChecked(false);
        }
    }

    @Override
    public void onPlayClick() {
        LrcControlView.OnKaraokeEventListener.super.onPlayClick();

        mViewModel.musicToggleStart();
    }


    @Override
    public void onStartSing() {
        mViewModel.leaveChorus();
    }

    @Override
    public void onJoinChorus() {
        mViewModel.joinChorus();
    }

    @Override
    public void onLeaveChorus() {
        mViewModel.leaveChorus();
    }

    @Override
    public void onDragTo(long time) {
        mViewModel.musicSeek(time);
    }

    @Override
    public void onLineFinished(LyricsLineModel line, int score, int cumulativeScore, int index, int total) {
        mLrcControlView.updateScore(score, cumulativeScore, /** Workaround(Hai_Guo)*/total * 100);
        mViewModel.syncSingleLineScore(score, cumulativeScore, index, total * 100);
    }

    @Override
    public void onSkipPreludeClick() {
        LyricsModel lyrics = mLrcControlView.getKaraokeView().getLyricsData();
        if (lyrics == null) {
            return;
        }
        // Experience will be better when seeking 2000 milliseconds ahead
        long seekPosition = lyrics.startOfVerse - 2000;
        mViewModel.musicSeek(seekPosition > 0 ? seekPosition : 0);
    }

    @Override
    public void onSkipPostludeClick() {
        LyricsModel lyrics = mLrcControlView.getKaraokeView().getLyricsData();
        if (lyrics == null) {
            return;
        }
        mViewModel.musicSeek(mViewModel.getSongDuration() - 500);
    }

    @Override
    public void onReGetLrcUrl() {
        mViewModel.reGetLrcUrl();
    }
}

File: Android/scenes/cantata/src/main/java/io/agora/scene/cantata/ui/widget/LrcActionListenerImpl.kt
package io.agora.scene.cantata.ui.widget

import android.content.Context
import io.agora.karaoke_view.v11.model.LyricsLineModel
import io.agora.scene.cantata.ui.activity.RoomLivingActivity
import io.agora.scene.cantata.ui.viewmodel.RoomLivingViewModel
import io.agora.scene.cantata.ui.widget.lrcView.LrcControlView

open class LrcActionListenerImpl constructor(
    private val mContext: Context,
    private val mViewModel: RoomLivingViewModel,
    private val mLrcControlView: LrcControlView
) : LrcControlView.OnKaraokeEventListener {
    override fun onSwitchOriginalClick() {
        super.onSwitchOriginalClick()
        mViewModel.musicToggleOriginal()
        mLrcControlView.setSwitchOriginalChecked(mViewModel.isOriginalMode())
    }

    override fun onPlayClick() {
        mViewModel.musicToggleStart()
    }

    override fun onStartSing() {
        mViewModel.leaveChorus()
    }

    override fun onJoinChorus() {
        if (mContext is RoomLivingActivity) {
            mContext.toggleSelfAudio(true, callback = {
                mViewModel.joinChorus()
            })
        }
    }

    override fun onLeaveChorus() {
        mViewModel.leaveChorus()
    }

    override fun onDragTo(time: Long) {
        mViewModel.musicSeek(time)
    }

    override fun onLineFinished(line: LyricsLineModel?, score: Int, cumulativeScore: Int, index: Int, total: Int) {
        mViewModel.updateSeatScoreStatus(score, cumulativeScore)
    }

    override fun onSkipPreludeClick() {
        val lyrics = mLrcControlView.karaokeView!!.lyricsData ?: return
        // Experience will be better when seeking 2000 milliseconds ahead
        val seekPosition = lyrics.startOfVerse - 2000
        mViewModel.musicSeek(if (seekPosition > 0) seekPosition else 0)
    }

    override fun onSkipPostludeClick() {
        val lyrics = mLrcControlView.karaokeView!!.lyricsData ?: return
        mViewModel.musicSeek(mViewModel.getSongDuration()!! - 500)
    }

    override fun onReGetLrcUrl() {
        mViewModel.reGetLrcUrl()
    }
}